app/src/test/java/com/codesoom/assignment/application/ProductServiceTest.java

```diff
@@ -0,0 +1,159 @@
+package com.codesoom.assignment.application;
+
+import com.codesoom.assignment.ProductNotFoundException;
+import com.codesoom.assignment.domain.Product;
+import com.codesoom.assignment.repository.ProductRepository;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+@DisplayName("ProductService 클래스 테스트")
+class ProductServiceTest {
+
+    private ProductServiceImpl service;
+    private ProductRepository repository;
+
+    private static final String PRODUCT_NAME = "춘식이 고구마 장난감";
+    private static final String MAKER        = "카카오";
+    private static final Long PRICE          = 39000L;
+    private static final String IMAGE_URL    = "http://localhost:8080/snake";
+
+    @BeforeEach
+    void setUp() {
+        //TODO 왜 mock로 감싸야하는지 강의 다시보기!
+        repository = mock(ProductRepository.class);
+        service = new ProductServiceImpl(repository);
+
+        List<Product> products = new ArrayList<>();
+        Product product = new Product(PRODUCT_NAME, MAKER, PRICE, IMAGE_URL);
+        products.add(product);
+
+        given(service.findAll()).willReturn(products);
+        given(service.findById(1L)).willReturn(Optional.of(product));
+        given(service.findById(100L)).willReturn(Optional.empty());
+    }
+
+    @Nested
+    @DisplayName("save 메소드는")
+    class Describe_save {
+        @Nested
+        @DisplayName("id가 존재하지 않으면")
+        class Context_without_id {
+            @Test
+            @DisplayName("product를 리포지토리에 저장한다")
+            void it_save_product() {
+                Product product = new Product(PRODUCT_NAME, MAKER, PRICE, IMAGE_URL);
+
+                service.save(product);
+
+                verify(repository).save(product);
+
+                assertThat(service.findById(1L)).isNotNull();
+                assertThat(service.findById(1L)
+                                    .get()
+                                    .getProductName())
+                                    .isEqualTo(PRODUCT_NAME);
+            }
+        }
+
+        @Nested
+        @DisplayName("id가 존재하면")
+        class Context_with_id {
+            @Test
+            @DisplayName("id에 해당하는 product를 업데이트한다")
+            void it_update_product() {
+                Product product = service.findById(1L)
+                                            .stream()
+                                            .findFirst()
+                                            .orElseThrow(ProductNotFoundException::new);
```

Comment:
`@BeforeEach`에서 `given`으로 상황을 명확하게 표현하니 `orElseThrow`대신 `get`을 사용해도 될까요?

Reply:
여기서는 `get`을 사용해도 괜찮겠어요. 정답은 있는 건 아니고, 어떤 의도를 드러내냐는 건데, 처음 들었던 생각은 테스트용 데이터가 잘 준비 안되있는 것을 여기서 예외를 던질 필요가 없다 라는 것이었습니다.

> If a value is present, returns the value, otherwise throws {@code NoSuchElementException}.

위의 내용은 고 `Optional.get`의 Javadocs에 써있는 내용인데, `NoSuchElementException`가 던져지므로 반대 의견도 있을 수 있는데, 만약 여기서 get을 했을 때 `ProductNotFoundException`예외가 아니라 다른 예외가 발생할테니 더 안좋다고 볼 수 있겠네요